File: go/v2/state/commitment.go
// Package state 提供 AUN E2EE V2 群状态承诺（state_commitment）计算。
//
// 规范引用 §6.2：
//
//	state_commitment = SHA256(
//	  "AUN-V2-SC-v1" ||
//	  group_aid ||
//	  uint32_BE(epoch) ||
//	  canonical_json(payload)
//	)
//
// payload 在序列化前需要内部规范化排序：
//   - members 按 aid 升序，每个 member 的 devices 按 device_id 升序
//   - audit_aids 升序
//   - admin_set.admin_aids 升序
//   - recovery_quorum.quorum_aids 升序
//
// 调用方不需要预排序，函数内部完成深拷贝 + 排序，避免修改原数据。
package state

import (
	"crypto/sha256"
	"encoding/binary"
	"encoding/hex"
	"sort"
	"strings"

	"github.com/modelunion/aun-sdk-core/go/v2/crypto"
)

// Prefix state_commitment 域分隔前缀
const Prefix = "AUN-V2-SC-v1"

// ComputeStateCommitment 计算群 state_commitment（hex 字符串）。
//
// statePayload 通常包含字段：members / audit_aids / join_policy_hash /
// admin_set / recovery_quorum / history_policy / wrap_protocol。
//
// 函数内部对 payload 做深拷贝并排序，输入数据不会被修改。
func ComputeStateCommitment(groupID string, epoch uint32, statePayload map[string]any) string {
	payload := deepCopyAny(statePayload).(map[string]any)
	sortPayload(payload)
	groupKey := normalizeGroupKey(groupID)

	var epochBytes [4]byte
	binary.BigEndian.PutUint32(epochBytes[:], epoch)

	canonical := crypto.CanonicalJSON(payload)

	h := sha256.New()
	h.Write([]byte(Prefix))
	h.Write([]byte(groupKey))
	h.Write(epochBytes[:])
	h.Write(canonical)
	return hex.EncodeToString(h.Sum(nil))
}

func normalizeGroupKey(groupID string) string {
	value := strings.ToLower(strings.Trim(strings.TrimSpace(groupID), "/"))
	if value == "" {
		return ""
	}
	trimDots := func(s string) string {
		return strings.Trim(s, ".")
	}
	if strings.HasPrefix(value, "group.") && strings.Contains(value, "/") {
		issuerAndBase := value[6:]
		if slash := strings.Index(issuerAndBase, "/"); slash > 0 && slash < len(issuerAndBase)-1 {
			domain := trimDots(issuerAndBase[:slash])
			baseTail := strings.Trim(issuerAndBase[slash+1:], "/")
			if at := strings.Index(baseTail, "@"); at > 0 {
				base := trimDots(baseTail[:at])
				suffixDomain := trimDots(baseTail[at+1:])
				if base != "" && suffixDomain != "" {
					merged := suffixDomain
					if domain != "" {
						merged = suffixDomain + "." + domain
					}
					return base + "." + merged
				}
			}
			base := trimDots(baseTail)
			if base != "" && domain != "" {
				return base + "." + domain
			}
		}
		return value
	}
	if at := strings.Index(value, "@"); at > 0 {
		base := trimDots(value[:at])
		domain := trimDots(value[at+1:])
		if base != "" && domain != "" {
			return base + "." + domain
		}
	}
	return value
}

// sortPayload 在原 map 上递归排序需要规范化的字段。
func sortPayload(payload map[string]any) {
	if members, ok := payload["members"].([]any); ok {
		sort.SliceStable(members, func(i, j int) bool {
			return memberAID(members[i]) < memberAID(members[j])
		})
		for _, m := range members {
			memberMap, ok := m.(map[string]any)
			if !ok {
				continue
			}
			if devices, ok := memberMap["devices"].([]any); ok {
				sort.SliceStable(devices, func(i, j int) bool {
					return deviceID(devices[i]) < deviceID(devices[j])
				})
			}
		}
	}

	sortStringList(payload, "audit_aids")

	if adminSet, ok := payload["admin_set"].(map[string]any); ok {
		sortStringList(adminSet, "admin_aids")
	}

	if quorum, ok := payload["recovery_quorum"].(map[string]any); ok {
		sortStringList(quorum, "quorum_aids")
	}
}

func memberAID(v any) string {
	if m, ok := v.(map[string]any); ok {
		if s, ok := m["aid"].(string); ok {
			return s
		}
	}
	return ""
}

func deviceID(v any) string {
	if m, ok := v.(map[string]any); ok {
		if s, ok := m["device_id"].(string); ok {
			return s
		}
	}
	return ""
}

func sortStringList(parent map[string]any, key string) {
	list, ok := parent[key].([]any)
	if !ok {
		return
	}
	sort.SliceStable(list, func(i, j int) bool {
		si, _ := list[i].(string)
		sj, _ := list[j].(string)
		return si < sj
	})
}

// deepCopyAny 对任意（JSON 解析得到的）值做深拷贝，避免排序修改调用方数据。
func deepCopyAny(v any) any {
	switch t := v.(type) {
	case map[string]any:
		out := make(map[string]any, len(t))
		for k, val := range t {
			out[k] = deepCopyAny(val)
		}
		return out
	case []any:
		out := make([]any, len(t))
		for i, item := range t {
			out[i] = deepCopyAny(item)
		}
		return out
	default:
		// 基础值（string/bool/json.Number/float64/nil）按值复制即可
		return v
	}
}

File: ts/src/v2/state/commitment.ts
/**
 * AUN E2EE V2: state_commitment 计算
 *
 * 规范引用：§6.2
 *
 * state_commitment = SHA256(
 *   "AUN-V2-SC-v1" ||
 *   group_aid ||
 *   uint32(epoch, big-endian) ||
 *   canonical_json({
 *     "members": [...sorted by aid, devices sorted by device_id...],
 *     "audit_aids": [...sorted...],
 *     "join_policy_hash": "64hex" | null,
 *     "admin_set": {"admin_aids": [...sorted...], "threshold": N},
 *     "recovery_quorum": {...} | null,
 *     "history_policy": "none" | "recent_N_days" | "full",
 *     "wrap_protocol": "3DH" | "1DH"
 *   })
 * )
 *
 * 排序在内部完成，调用方无需预排序。
 */

import { createHash } from 'node:crypto';
import { normalizeGroupId } from '../../group-id.js';
import { canonicalJson } from '../crypto/canonical.js';

export const STATE_PREFIX: Uint8Array = new TextEncoder().encode('AUN-V2-SC-v1');

export interface MemberDevice {
  device_id?: string;
  fp?: string;
  [k: string]: unknown;
}
export interface Member {
  aid?: string;
  devices?: MemberDevice[];
  [k: string]: unknown;
}
export interface AdminSet {
  admin_aids?: string[];
  threshold?: number;
  [k: string]: unknown;
}
export interface RecoveryQuorum {
  trigger?: string;
  quorum_aids?: string[];
  threshold?: number;
  [k: string]: unknown;
}
export interface StatePayload {
  members?: Member[];
  audit_aids?: string[];
  join_policy_hash?: string | null;
  admin_set?: AdminSet;
  recovery_quorum?: RecoveryQuorum | null;
  history_policy?: string;
  wrap_protocol?: string;
  [k: string]: unknown;
}

/** in-place 规范化排序。 */
function sortPayload(payload: StatePayload): void {
  if (Array.isArray(payload.members)) {
    payload.members.sort((a, b) => {
      const ka = a.aid ?? '';
      const kb = b.aid ?? '';
      if (ka === kb) return 0;
      return ka < kb ? -1 : 1;
    });
    for (const m of payload.members) {
      if (Array.isArray(m.devices)) {
        m.devices.sort((a, b) => {
          const ka = a.device_id ?? '';
          const kb = b.device_id ?? '';
          if (ka === kb) return 0;
          return ka < kb ? -1 : 1;
        });
      }
    }
  }
  if (Array.isArray(payload.audit_aids)) {
    payload.audit_aids.sort();
  }
  if (payload.admin_set && Array.isArray(payload.admin_set.admin_aids)) {
    payload.admin_set.admin_aids.sort();
  }
  if (payload.recovery_quorum && Array.isArray(payload.recovery_quorum.quorum_aids)) {
    payload.recovery_quorum.quorum_aids.sort();
  }
}

/** 大端 uint32 编码。 */
function uint32BE(value: number): Uint8Array {
  if (value < 0 || value > 0xffffffff || !Number.isInteger(value)) {
    throw new Error(`epoch out of uint32 range: ${value}`);
  }
  const out = new Uint8Array(4);
  const dv = new DataView(out.buffer);
  dv.setUint32(0, value, false);
  return out;
}

/**
 * 计算 state_commitment（hex）。
 */
export function computeStateCommitment(
  groupId: string,
  epoch: number,
  statePayload: StatePayload,
): string {
  // 深拷贝避免修改调用方数据
  const payload: StatePayload = JSON.parse(JSON.stringify(statePayload));
  sortPayload(payload);

  const groupKey = normalizeGroupId(groupId) || String(groupId ?? '').trim();
  const groupBytes = new TextEncoder().encode(groupKey);
  const epochBytes = uint32BE(epoch);
  const payloadBytes = canonicalJson(payload);

  const h = createHash('sha256');
  h.update(STATE_PREFIX);
  h.update(groupBytes);
  h.update(epochBytes);
  h.update(payloadBytes);
  return h.digest('hex');
}
